### PR TITLE
Tree estimators

### PR DESCRIPTION
**Pull request by sfilipi**
Ongoing work on converting the trainers to estimators. This PR converts the Tree -type Predictors.

### CODE REVIEW

File: src/Microsoft.ML.FastTree/FastTreeRanking.cs [hunk 1]
```diff
@@ -405,6 +414,9 @@ protected override string GetTestGraphHeader()
             return headerBuilder.ToString();
         }
 
+        protected override RankingPredictionTransformer<FastTreeRankingPredictor> MakeTransformer(FastTreeRankingPredictor model, ISchema trainSchema)
+        => new RankingPredictionTransformer<FastTreeRankingPredictor>(Host, model, trainSchema, FeatureColumn.Name);
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTreeRanking.cs]: >FeatureColumn.Name); [](start = 96, length = 20)

should add the GroupID to the base constructor

Comment (Zruty0) [reply src/Microsoft.ML.FastTree/FastTreeRanking.cs]: GroupID? why?

---
In reply to: [216093781](https://github.com/dotnet/machinelearning/pull/855#discussion_r216093781) [](ancestors = 216093781)

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 2]
```diff
@@ -133,6 +136,18 @@ protected virtual Float GetMaxLabel()
             return Float.PositiveInfinity;
         }
 
+        private static SchemaShape.Column MakeWeightColumn(Optional<string> weightColumn)
+        {
+            if (weightColumn == null || !weightColumn.IsExplicit)
```
[the remark this answers is not in the document]
Comment (Zruty0) [reply src/Microsoft.ML.FastTree/FastTree.cs]: we cannot do this really, can we?

---
In reply to: [216135820](https://github.com/dotnet/machinelearning/pull/855#discussion_r216135820) [](ancestors = 216135820)

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/SimpleTrainAndPredict.cs [hunk 3]
```diff
@@ -50,5 +51,48 @@ public void New_SimpleTrainAndPredict()
...
         }
+
+        /// <summary>
+        /// Start with a dataset in a text file. Run text featurization on text values. 
+        /// Train a linear model over that. (I am thinking sentiment classification.) 
+        /// Out of the result, produce some structure over which you can get predictions programmatically 
+        /// (e.g., the prediction does not happen over a file as it did during training).
+        /// </summary>
+        [Fact]
+        public void New_SimpleTrainAndPredictWithFT()
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/Api/Estimators/SimpleTrainAndPredict.cs]: >New_SimpleTrainAndPredictWithFT [](start = 20, length = 31)

move this test somewhere else

File: src/Microsoft.ML.FastTree/FastTreeRanking.cs [hunk 4]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.ML.Runtime.Internal.Utilities;
 using Microsoft.ML.Runtime.Model;
 using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Core.Data;
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/FastTreeRanking.cs]: >using [](start = 0, length = 5)

sort

File: src/Microsoft.ML.FastTree/FastTreeRanking.cs [hunk 5]
```diff
@@ -53,17 +55,25 @@ public sealed partial class FastTreeRankingTrainer : BoostingFastTreeTrainerBase
...
+
         public FastTreeRankingTrainer(IHostEnvironment env, Arguments args)
-                : base(env, args)
+                : base(env, args, MakeLabelColumn(args.LabelColumn))
         {
+            OutputColumns = new[]
+            {
+                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false),
+                new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false),
+                new SchemaShape.Column(DefaultColumnNames.PredictedLabel, SchemaShape.Column.VectorKind.Scalar, BoolType.Instance, false)
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTreeRanking.cs]: double-check this is correct

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 6]
```diff
@@ -87,34 +90,53 @@ public abstract class FastTreeTrainerBase<TArgs, TPredictor> :
...
+            //apply the advanced args, if the user supplied any
+            advancedSettings?.Invoke(Args);
+            Args.LabelColumn = label.Name;
+
+            if (weightColumn != null)
+                Args.WeightColumn = Optional<string>.Explicit(weightColumn);
+
+            if (groupIdColumn != null)
+                Args.GroupIdColumn = Optional<string>.Explicit(groupIdColumn);
+
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTree.cs]: @tfinley@gmail.com what is the desired behavior here, should we preserve the Optional Implicing strings, or the code will handle null?

File: src/Microsoft.ML.Ensemble/OutputCombiners/MultiStacking.cs [hunk 7]
```diff
@@ -57,7 +57,7 @@ public Arguments()
                     env => new Ova(env, new Ova.Arguments()
                     {
                         PredictorType = ComponentFactoryUtils.CreateFromFunction(
-                             e => new AveragedPerceptronTrainer(e, new AveragedPerceptronTrainer.Arguments()))
+                            e => new FastTreeBinaryClassificationTrainer(e, DefaultColumnNames.Label, DefaultColumnNames.Features))
```
Comment (TomFinley) [inline src/Microsoft.ML.Ensemble/OutputCombiners/MultiStacking.cs]: >FastTreeBinaryClassificationTrainer [](start = 37, length = 35)

I'd really rather we didn't. This seems to fit into the same bucket as the discussion on #682. That ensembling should have a dependency on FastTree merely because we have a default does not make sense to me. If someone wants to use stacking, that's great, but they need to specify the learners.

Comment (TomFinley) [reply src/Microsoft.ML.Ensemble/OutputCombiners/MultiStacking.cs]: But maybe we can hold off for right now.

---
In reply to: [218215145](https://github.com/dotnet/machinelearning/pull/855#discussion_r218215145) [](ancestors = 218215145)

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 8]
```diff
@@ -25,6 +25,8 @@
 using Microsoft.ML.Runtime.Training;
 using Microsoft.ML.Runtime.TreePredictor;
 using Newtonsoft.Json.Linq;
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime.EntryPoints;
```
Comment (TomFinley) [inline src/Microsoft.ML.FastTree/FastTree.cs]: I'm probably just missing something obvious, but why does this now depend on entry-points namespace?

Also sorting.

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs [hunk 9]
```diff
@@ -301,6 +301,52 @@ private static VersionInfo GetVersionInfo()
         }
     }
 
+    public sealed class RankingPredictionTransformer<TModel> : PredictionTransformerBase<TModel>
```
[the remark this answers is not in the document]
Comment (Zruty0) [reply src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: I think this transformer should also expose the group ID column, at least that would be my belief

---
In reply to: [218214277](https://github.com/dotnet/machinelearning/pull/855#discussion_r218214277) [](ancestors = 218214277)

[the remark this answers is not in the document]
Comment (TomFinley) [reply src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: Yes good point, if the others expose the label and features then this ought to expose the group ID as well.

---
In reply to: [218216192](https://github.com/dotnet/machinelearning/pull/855#discussion_r218216192) [](ancestors = 218216192,218214277)

**Comment by TomFinley** [comment]
Is omission of Pigsty extensions deliberate?

File: src/Microsoft.ML.Ensemble/OutputCombiners/MultiStacking.cs (hunk 7 above)
Comment (sfilipi) [reply src/Microsoft.ML.Ensemble/OutputCombiners/MultiStacking.cs]: Yes, let's do that separately, when we shape the ensembles. 

---
In reply to: [218215323](https://github.com/dotnet/machinelearning/pull/855#discussion_r218215323) [](ancestors = 218215323,218215145)

**Comment by sfilipi** [comment]
Did i misunderstand that for trainers we should hold on to doing the Pigsty extensions until we get the ml task, so we could extend on that, rather than the label? @tfinley@gmail.com @Zruty0, let me know if i should actually work on them in the same PR.  

---
In reply to: [422161730](https://github.com/dotnet/machinelearning/pull/855#issuecomment-422161730) [](ancestors = 422161730)

**Comment by Zruty0** [comment]
I have the same (mis)understanding. In any case, let's do it after this oner

---
In reply to: [422174998](https://github.com/dotnet/machinelearning/pull/855#issuecomment-422174998) [](ancestors = 422174998,422161730)

File: src/Microsoft.ML.FastTree/FastTree.cs (hunk 8 above)
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTree.cs]: Thank you! Oversight

---
In reply to: [218216150](https://github.com/dotnet/machinelearning/pull/855#discussion_r218216150) [](ancestors = 218216150)
